crates/wasi-common/tokio/src/sched/unix.rs

```diff
@@ -0,0 +1,90 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll as FPoll};
+use wasi_common::{
+    sched::{
+        subscription::{RwEventFlags, Subscription},
+        Poll,
+    },
+    Context as _, Error,
+};
+
+struct FirstReady<'a, T>(Vec<Pin<Box<dyn Future<Output = T> + 'a>>>);
+
+impl<'a, T> FirstReady<'a, T> {
+    fn new() -> Self {
+        FirstReady(Vec::new())
+    }
+    fn push(&mut self, f: impl Future<Output = T> + 'a) {
+        self.0.push(Box::pin(f));
+    }
+}
+
+impl<'a, T> Future for FirstReady<'a, T> {
+    type Output = T;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> FPoll<T> {
+        let mut result = FPoll::Pending;
+        for f in self.as_mut().0.iter_mut() {
+            match f.as_mut().poll(cx) {
+                FPoll::Ready(r) => match result {
+                    // First ready gets to set the result. But, continue the loop so all futures
+                    // get the opportunity to become ready.
```

Comment:
How come this continues polling all futures rather than just returning the first ready value immediately?

Reply:
In the corner case where several of your subscriptions are ready immediately, the tests expect all of them to be reported rather than just the first. Reporting is part of the future being polled on here, so I want all of them to get a chance to run before returning. I will improve the comment.